File: rutgers_projs/tbf_proj/intopost.c
// Convert infix to postfic notation

/*
  $Log$
*/

#include <ctype.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "mapping.h"
#include "tbf_hdr.h"

//tries to get the best representation for the given number
void best_rep(char **postfix, float val) {
  char cval = val;
  short sval = val;

  // try char first
  if (cval == val) {
    *(*postfix)++ = CHAR;
    *((char *)(*postfix))++ = val;
  } else if (sval == val) {
    *(*postfix)++ = INT;
    *((short *)(*postfix))++ = val;
  } else {
    *(*postfix)++ = FLOAT;
    *((float *)(*postfix))++ = val;
  }
}

// printout the given packet
int bin_print(const char *postfix_str, int k) {
  int lcv;
  for(lcv = 0; lcv < k; lcv ++)
    printf("%02x ", postfix_str[lcv] & 0xff);
  printf("\n");
  fflush(stdout);
  return lcv;
}

// printout the given packet in human readable form
int symb_print(const char *str, int len)
{
  unsigned char c;
  const char *orig = str;
  
  while (len > (str - orig)) {
    switch (c = *str++) {
    case FLOAT:
      printf("%lf ", ((double)*((float*)str)++));
      break;

      // i think the motes have 2byte ints
      // we definitely need to fix this
      // FIXME:
    case INT:
      printf("%lf ",  ((double)*((short*)str)++));
      break;

    case CHAR:
      printf("%lf ", ((double)*((char*)str)++));
      break;

    case T: {
      printf("t ");
      break;
    }
    case MY_PI:
      printf("PI ");
      break;

    default:
	printf("%s ", s_table[c]);
    } // switch
  } // while
  printf("\n");
  return 1;

}

 /**************************************************************\
 |** Mike:  Checks lookup table for alphas.  Returns location **|
 |**        if found else returns -1                          **|
 \**************************************************************/
int lookup(const char* w)
{
  int lcv = 0;

  // change +-/*
  switch (*w) {
  case '+':
    w = "my_add";
    break;
  case '*':
    w = "my_mul";
    break;
  case '/':
    w = "my_div";
    break;
  case '-':
    w = "my_sub";
    break;
  case '^':
    w = "pow";
    break;
  }

  while ((lcv < TABLE_SIZE) && (strcmp(s_table[lcv],w) != 0))
    lcv++;
  if (lcv+1 > TABLE_SIZE) {
    return (-1);
  }
  else {
    return (lcv);
  }
}

// given two infix strings, and a pointer to tbf_hdr,
// encodes the infix in postfix form and dumps it
// into the packet
// returns -1 on failure
inline int dump_traj(const char *inf_x, const char *inf_y,
		     struct tbf_hdr *ptbf_hdr, int data_len)
{
  char *x_traj;
  char *y_traj;
  int err;

  // do x
  x_traj = (char *) (ptbf_hdr + 1);
  err = intopost(inf_x, x_traj);
  if (err < 0) return -1;
  else if (err + sizeof(struct tbf_hdr) > data_len) {
#ifdef TBF_DEBUG
    printf("x String too big\n");
#endif
    return -2;
  }
  ptbf_hdr->x_byte = err;

  // do y
  y_traj = (char *) (x_traj + (int)ptbf_hdr->x_byte);
  err = intopost(inf_y, y_traj);
  if (err < 0) return -1;
  else if (err + ptbf_hdr->x_byte + sizeof(struct tbf_hdr) > data_len) {
#ifdef TBF_DEBUG
    printf("y String too big\n");
#endif
    return -2;
  }
  ptbf_hdr->y_byte = err;

#ifdef TBF_DEBUG
  printf("encodings: \n");
  symb_print(x_traj, ptbf_hdr->x_byte);
  symb_print(y_traj, ptbf_hdr->y_byte);
#endif

  return 1;
}


/*--------------------------------------------------------------------------

FUNCTION: intopost - converts infix to postfix

A.	Push a ( on the stack. This sentinel allows us to detect when we
	have flushed out the stack on completion in step 1.

--- Perform steps B through H for each character in the infix string ---

B.	Ignore white space.

C.	Pass alphabetic characters through to postfix list.

D.	Push any ( on the stack. These sentinels allows us to detect when
	have flushed out the stack when handling ) and operators.

E.	Have a ) so pop off the stack and put into postfix list until a (
	is popped. Discard the (.

F	Have a * or /. Pop off any operators of equal or higher precedence
	and put them into postfix list. If a ( or lower precedence operator
	(such as + or -) is popped, put it back and stop looking.
	Push new * or /.

G.	Have a + or -. Pop off any operators of equal or higher precedence
	(that includes all of them) and put them into postfix list. If a (
	is popped, put it back and stop looking. Push new + or -.

H.	Report unknown character on input.

--------

I.	Have processed all input characters. New flush stack until we find
	the matchint ( put there in step A.

J.	Terminate the postfix list.

--------------------------------------------------------------------------*/

int intopost(const char *infix, char *postfix)
{
    int st;
    char *orig = postfix; 
    struct stack istack;

    stack_init(&istack);

    /*A*/
    push(&istack, '(');
    while (*infix != '\0') {
      
#ifdef TRACE
      printf("*infix: %c\n", *infix);
#endif

      /*B*/
      if (isspace(*infix)) {
	  ;
      } else if (isdigit(*infix)) {
	/*C*/
	float a;
	int b;
	// read the number as a float
	sscanf(infix, "%f%n", &a, &b);
	best_rep(&postfix, a);
	infix += b-1;

      } else if (isalpha(*infix)) {
	// scans funcs 
	int b, code;
	char w[256];
	sscanf(infix, "%[_a-zA-Z]%n", w, &b);
	// see if it is a t
	if (strcasecmp(w, "t") == 0) {
	  *postfix++ = T;
	} else if (strcasecmp(w, "pi") == 0) {
	  *postfix++ = MY_PI;
	} else {
	  code = lookup(w);
	  if (code == -1) {
	    printf("%s is not valid code\n", w);
	    return -1;
	  }
	  push(&istack, code);
	}
	infix += b-1;

      } else if (*infix == '(') {
	/*D*/
	push(&istack, '(');

      } else if (*infix == ')') {
	/*E*/
	while ((st = pop(&istack)) != '(') {
	  *postfix++ = st;
	}
      } else if (*infix == '*' || *infix == '/' || *infix == '^') {
	char str[] = {*infix};
	int code = lookup(str);
	/*F*/
	while (1) {
	  if ((st = pop(&istack)) == '(' || st == '+' || st == '-') {
	    push(&istack, st);
	    break;
	  }
	  *postfix++ = st;
	}
	if (code != -1) 
	  push(&istack, code);
	else {
	  printf("Unknown symbol %s\n", str);
	  return -1;
	}
	
      } else if (*infix == '+' || *infix == '-') {
	char str[] = {*infix};
	int code = lookup(str);
	/*G*/
	while (1) {
	  if ((st = pop(&istack)) == '(') {
	    push(&istack, st);
	    break;
	  }
	  *postfix++ = st;
	}
	if (code != -1) 
	  push(&istack, code);
	else {
	  printf("Unknown symbol %s\n", str);
	  return -1;
	}
      } else {
	/*H*/
	printf("Unknown input character %c\n", *infix);
	return -1;
      }
      ++infix;
      continue;
    }
    /*I*/
    while ((st = pop(&istack)) != '(') {
      *postfix++ = st;
      
    }

    /*J*/
    *postfix = '\0';
    return (postfix - orig);
}

File: rutgers_projs/tbf_proj/java_ui/intopost.java
/*
	INTOPOST.JAVA
	Ported to java from C
*/
package tbf.java_ui;

import java.util.*;

public class intopost{

String FLOAT= "01";//     = 1;
String INT= "02";//       = 2;
String CHAR="03";//	= 3;
String SIN="04";//        = 4;
String COS="05";//        = 5;
String TAN="06";//        = 6;
String ASIN="07";//       = 7;
String ACOS="08";//       = 8;
String ATAN="09";//       = 9;
String SINH="0a";//       = 10;
String COSH="0b";//       = 11;
String TANH="0c";//       = 12;
String ASINH="0d";//      = 13;
String ACOSH="0e";//	= 14;
String ATANH="0f";//	= 15;
String LOG10="10";//	= 16;
String LOG="11";//	= 17;
String EXP="12";//	= 18;
String SQRT="13";//	= 19;
String CEIL="14";//	= 20;
String FLOOR="15";//	= 21;
String MY_PI="16";//	= 22;
String ADD="81";//	= 129;
String SUB="82";//	= 130;
String MUL="83";//	= 131;
String DIV="84";//	= 132;
String POW="85";//	= 133;
String T	="ff";//        = 255;
int TABLE_SIZE= 255;


  String s_table[] = {"NULL","NULL","NULL","NULL","sin","cos","tan","asin","acos","atan","sinh","cosh","tanh","asinh","acosh","atanh","log10","log","exp","sqrt","ceil","floor","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","my_add","my_sub","my_mul","my_div","pow","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL","NULL",};


    StringBuffer infix = null;
    StringBuffer postfix = null;
    String finished = null;
    Stack stk = null;
    int i_len;

    /*
     * Constructor
     */
    public intopost(String inf){
	infix = new StringBuffer(inf);
	postfix = new StringBuffer();
	i_len = inf.length();
	stk = new Stack();
	process();
    };

    public String getPostfix(){ 
	if(process()!=-1)
	    return finished;
	else
	    return null;
    }

    void best_rep(int pptr, float val) {
	char cval = (char)val;
	short sval = (short)val;

	System.out.println("got " + val);
  
	// try char first
	if (cval == val) {
	    System.out.println("putting as char\n");
	    postfix.append(CHAR);
	    postfix.append((char)val);
	} else if (sval == val) {
	    System.out.println("putting as short\n");
	    postfix.append(INT);
	    postfix.append((int)val);
	} else {
	    System.out.println("putting as float\n");
	    postfix.append(FLOAT);
	    postfix.append((float)val);
	}
    }

/*--------------------------------------------------------------------------

FUNCTION: intopost - converts infix to postfix

A.	Push a ( on the stack. This sentinel allows us to detect when we
	have flushed out the stack on completion in step 1.

--- Perform steps B through H for each character in the infix string ---

B.	Ignore white space.

C.	Pass alphabetic characters through to postfix list.

D.	Push any ( on the stack. These sentinels allows us to detect when
	have flushed out the stack when handling ) and operators.

E.	Have a ) so pop off the stack and put into postfix list until a (
	is popped. Discard the (.

F	Have a * or /. Pop off any operators of equal or higher precedence
	and put them into postfix list. If a ( or lower precedence operator
	(such as + or -) is popped, put it back and stop looking.
	Push new * or /.

G.	Have a + or -. Pop off any operators of equal or higher precedence
	(that includes all of them) and put them into postfix list. If a (
	is popped, put it back and stop looking. Push new + or -.

H.	Report unknown character on input.

--------

I.	Have processed all input characters. New flush stack until we find
	the matchint ( put there in step A.

J.	Terminate the postfix list.

--------------------------------------------------------------------------*/
    int parseFloat( StringBuffer sb, int start, int end, float f){
	StringBuffer ret = new StringBuffer();
	int x;

	for(x=start;x<end;x++){
	    if(Character.isDigit(sb.charAt(x))||sb.charAt(x)=='.'){
		ret.append(sb.charAt(x));
	    }
	    else
		break;
	}

	f=Float.parseFloat(ret.toString());
	return x;
    }

    int parseString( StringBuffer sb, int start, int end, String s){
	StringBuffer ret = new StringBuffer();
	int x;

	for(x=start;x<end;x++){
	    if(Character.isLetter(sb.charAt(x))||
	       (!Character.isLetterOrDigit(sb.charAt(x))&&Character.isWhitespace(sb.charAt(x)))){
		ret.append(sb.charAt(x));
	    }
	    else
		break;
	}

	s=ret.toString();
	return x;
    }

    int process()
    {

	String st;
	int ptr_pf_orig = 0;
	int ptr_infix = 0;
	int ptr_postfix = 0;

	/*A*/
	stk.push("(");
	while (ptr_infix != i_len) {
	    

	    /*B*/
	    if (Character.isSpaceChar(infix.charAt(ptr_infix))) {
		;
	    } else if (Character.isDigit(infix.charAt(ptr_infix))) {
		/*C*/
		float a = 0;
		int b = 0;
		// read the number as a float
		b=parseFloat(infix, ptr_infix, i_len, a);
		best_rep(ptr_postfix, a);
		ptr_infix = b;
 System.out.println("Saw a number "+a);		
	    } else if (Character.isLetter(infix.charAt(ptr_infix))) {
		// scans funcs 
		int b=0;
		String w = null;
		b = parseString(infix,ptr_infix,i_len,w);
 System.out.println("Saw a word "+w);
		System.out.println("Got "+w+", consumed "+ b);
		// see if it is a t
		if (w.compareToIgnoreCase("t")==0) {
		    postfix.append(T);
		} else if (w.compareToIgnoreCase("pi")==0 ) {
		    postfix.append(MY_PI);
		} else {
		    String code = lookup(w);
		    if (code.compareTo("-1")==0) {
			System.out.println(w+" is not valid code\n");
			return -1;
		    }
		    stk.push(code);
		}
		ptr_infix = b;
		
	    } else if (infix.charAt(ptr_infix) == '(') {
		/*D*/
		stk.push("(");
		ptr_infix++;
		
	    } else if (infix.charAt(ptr_infix) == ')') {
		/*E*/
		while ((st = (String)stk.pop()).compareTo("(")==0) {
		    postfix.append(st);		
		}
		ptr_infix++;
	    } else if (infix.charAt(ptr_infix) == '*' || infix.charAt(ptr_infix) == '/' || 
		       infix.charAt(ptr_infix) == '^') {
		char[] tmp = {infix.charAt(ptr_infix)};
		String str = new String(tmp);
 System.out.println("Saw this op: "+str);	
		String code = lookup(str);
		/*F*/
		while (true) {
		    if ((st = (String)stk.pop()).compareTo("(") == 0 || st.compareTo("+") == 0
			|| st.compareTo("-") == 0) {
			stk.push(st);
			break;
		    }
		    postfix.append(st);
		}
		if (code.compareTo("-1")!=0) 
		    stk.push(code);
		else {
		    System.out.println("Unknown symbol " + str);
		    return -1;
		}
		ptr_infix++;
	    } else if (infix.charAt(ptr_infix) == '+' || infix.charAt(ptr_infix) == '-') {
		char tmp[] = {infix.charAt(ptr_infix)};
		String str = new String(tmp);
		String code = lookup(str);
 System.out.println("Saw this op: "+str);
		/*G*/
		while (true) {
		    if ((st = (String)stk.pop()).compareTo("(") == 0) {
			stk.push(st);
			break;
		    }
		    postfix.append(st);
		}
		if (code.compareTo("-1") != 0) 
		    stk.push(code);
		else {
		    System.out.println("Unknown symbol" +  str);
		    return -1;
		}
		ptr_infix++;
	    } else {
		/*H*/
		System.out.println("Unknown input character "+ infix.charAt(ptr_infix));
	    }
	   
	    continue;
	}
	/*I*/
	while ((st =(String)stk.pop()).compareTo("(") != 0) {
	    postfix.append(st);
	    
	}
	
	/*J*/
       
        finished = postfix.toString();
	return 1;
    }

 /**************************************************************\
 |** Mike:  Checks lookup table for alphas.  Returns location **|
 |**        if found else returns -1                          **|
 \**************************************************************/
String lookup(String w)
{
  int lcv = 0;

  // change +-/*
  if(w.compareTo("+")==0)
    w = "my_add";
  else if(w.compareTo("*")==0)
    w = "my_mul";
  else if (w.compareTo("/")==0)
    w = "my_div";
  else if (w.compareTo("-")==0)
    w = "my_sub";
  else if (w.compareTo("^")==0)
    w = "pow";
 
  System.out.println("Got "+w+", Lookup says: ");

  while ((lcv < TABLE_SIZE) && ( s_table[lcv].compareTo(w)!= 0))
    lcv++;
  
  if (lcv+1 > TABLE_SIZE) {
    System.out.println("-1\n");
    return ("-1");
  }
  else {
    System.out.println(Integer.toHexString(lcv));
    if(lcv < 16)
	return ("0"+Integer.toHexString(lcv));
    return Integer.toHexString(lcv);
  }
}

double my_add(double a, double b) {return a+b;}
double my_mul(double a, double b) {return a*b;}
double my_div(double a, double b) {return a/b;}
double my_sub(double a, double b) {return a-b;}

/*
// returns a success or failure
int rpncalc(const char *str, int len, double t,
	    double *result) {
  unsigned char c;
  char *orig = str;


  System.out.println("rpncalc got len: %d\n", len);
  while (len > (str - orig)) {
    switch (c = *str++) {
    case FLOAT:
      push_dbl((double)*((float*)str)++);
      break;

      // i think the motes have 2byte ints
      // we definitely need to fix this
      // FIXME:
    case INT:
      push_dbl((double)*((short*)str)++);
      break;

    case CHAR:
      push_dbl((double)*((char*)str)++);
      break;

    case T: {
      double my_t = t;
      System.out.println("Got t, treating it as %lf\n", my_t);
      push_dbl(my_t);
      break;
    }
    case MY_PI:
      push_dbl(M_PI);
      break;

    default:
      if (c > START_ARG2_FUNCS) {
	// has 2 args
	double a;
	double b;
	double (*func)(double a, double b);

	a = pop_dbl();
	b = pop_dbl();
	// call the func
	func = f_table[c];	
	if (func != NULL) {
	  System.out.println("%s(%lf, %lf)\n", s_table[c], b, a);
	  push_dbl(func(b, a));
	} 
      } else {
	double (*func)(double a);
	double a = pop_dbl();

	// has only one arg
	// call the function
	func = f_table[c];
	if (func != NULL) {
	  System.out.println("%s(%lf)\n", s_table[c], a);
	  push_dbl(func(a));
	}
      }
      break;
    } // switch
  } // while
  *result = pop_dbl();
  if (stack_dbl_err() < 0) {
    clear_dbl();
    return (-1);
  }
  return 1;

  } // rpncalc*/
}
